Rest API returns custom error codes and Run_plugin endpoint returns a valid error when an incorrect graph_id is supplied.

Auriga2 opened the pull request:
<!--

### Requirements

* Filling out the template is required. Any pull request that does not include
enough information to be reviewed in a timely manner may be closed at the
maintainers' discretion.
* Follow the check list items defined by https://github.com/constellation-app/constellation/blob/master/CONTRIBUTING.md#pull-requests
* All new code requires unit tests to ensure they work as expected and will
continue to work as new code is added in the future (regression testing).
* Have you read Constellation's Code of Conduct? By filing an issue, you are
expected to comply with it, including treating everyone with respect:
https://github.com/constellation-app/constellation/blob/master/CODE_OF_CONDUCT.md

-->

### Description of the Change

* When any errors are thrown from the services, the Rest API was throwing them only as Error 500.
With this fix we can response with custom error codes.

* Run_plugin endpoint returns a valid error when an incorrect graph_id is supplied.
<!--

We must be able to understand the design of your change from this description.
If we can't get a good idea of what the code will be doing from the description
here, the pull request may be closed at the maintainers' discretion. Keep in
mind that the maintainer reviewing this PR may not be familiar with or have
worked with the code here recently, so please walk us through the concepts.

-->

### Alternate Designs
I managed to get it to work by passing the HttpServletResponse response object to the services classes. But tricky to refactor the FileListener.java class.
<!--

Explain what other alternates were considered and why the proposed version was
selected.

-->

### Why Should This Be In Core?
Better UX
<!--

Explain why this functionality should be in Constellation Core as opposed to a
different module suite. Note that this question is more applicable when adding
new functionality. If this change is a minor update to an exising file then it
is understood that this change has to be to this module suite and a response
and therefore a response to this question is not required.

-->

### Benefits
Better UX
<!-- What benefits will be realized by the code change? -->

### Possible Drawbacks
N/A
<!-- What are the possible side-effects or negative impacts of the code change? -->

### Verification Process
* In the Rest API, try the `SetGraph` endpoint with an invalid graph ID. It should return an Error 422 instead of 500
* Try the `RunPlugin` endpoint with an invalid graph ID. It should return an Error 422 instead of 200
<!--

What process did you follow to verify that your change has the desired effects?

- How did you verify that all new functionality works as expected?
- How did you verify that all changed functionality works as expected?
- How did you verify that the change has not introduced any regressions?

Describe the actions you performed (e.g., buttons you clicked, text you typed,
commands you ran, etc.), and describe the results you observed.

-->

### Applicable Issues

<!-- Link any applicable issues here -->
#577

arcturus2 commented on CoreWebServer/src/au/gov/asd/tac/constellation/webserver/restapi/RestService.java:
```diff
@@ -38,9 +38,56 @@
  * @author algol
  */
 public abstract class RestService {
+    public static final int SC_CONTINUE = 100;
```
Can you please use the `HttpURLConnection` class instead of creating these constants?

Auriga2 replied:
422 is not there. Shall I leave only that in the new class then?

arcturus2 replied:
Yep @Auriga2. Sorry I didn't realise 422 was not there.

arcturus2 commented on CoreWebServer/src/au/gov/asd/tac/constellation/webserver/services/AddRecordStore.java:
```diff
@@ -182,6 +183,7 @@ public void callService(final PluginParameters parameters, final InputStream in,
         }
 
         addToGraph(graphId, rs, completeWithSchema, arrange, resetView);
+        return new ServiceResponse(SC_OK, "Successful"); 
```
So in here it would be `HttpURLConnection.HTTP_OK`.
What is the reason to show the message "Successful" given the HTTP code will inform this. Does the message appear in the response output?
Also if you are going to keep it then its work making the string a constant.

Auriga2 replied:
@arcturus2 It's not really used afterwards, as I am processing only if it's not a 200. I just typed it thinking it's better than an empty string. What do you think?

arcturus2 replied:
If its not used then how about creating another constructor like this?

```
public ServiceResponse(int errorCode){
        this.errorCode = errorCode;
        this.errorMessage = null;
}
```

Auriga2 replied:
Agree, fixed.

arcturus2 commented on CoreWebServer/src/au/gov/asd/tac/constellation/webserver/restapi/ServiceResponse.java:
```diff
@@ -0,0 +1,46 @@
+ * callService in webserver.services classes.
+ */
+public class ServiceResponse { 
+    private String errorMessage;
```
how about `statusMessage`?

arcturus2 commented on CoreWebServer/src/au/gov/asd/tac/constellation/webserver/restapi/ServiceResponse.java:
```diff
@@ -0,0 +1,46 @@
+ */
+public class ServiceResponse { 
+    private String errorMessage;
+    private int errorCode;
```
how about `statusCode` given the response is not always when there is an error.

Auriga2 commented:
Fixed the above issues and also noticed that `set_graph` endpoint was still throwing 500 if no graph was loaded in Constellation. Added a check with `GraphNode.getAllGraphs().isEmpty()` but currently the PUT method doesn't show a custom message because it doesn't have a response body.
![image](https://user-images.githubusercontent.com/64826928/84992863-a3800980-b187-11ea-900b-68b6f1eff39f.png)

arcturus2 commented:
Thats awesome stuff @Auriga2. When you are ready please take it off draft and i will ask @algol60 to also do a review too.

arcturus2 commented on CoreWebServer/src/au/gov/asd/tac/constellation/webserver/api/RestServiceServlet.java:
```diff
@@ -101,7 +102,10 @@ private void callService(final HttpMethod httpMethod, final HttpServletRequest r
             response.setStatus(HttpServletResponse.SC_OK);
-            rs.callService(parameters, request.getInputStream(), response.getOutputStream());
+            ServiceResponse serviceResponse = rs.callService(parameters, request.getInputStream(), response.getOutputStream());
+            if (serviceResponse.getStatusCode()!=200){
```
@Auriga2  should the `200` be replaced with `HttpURLConnection.HTTP_OK`?

Auriga2 replied:
Yes Thanks! Fixed.

arcturus2 commented:
So @Auriga2 your change is pretty good. After discussing with @algol60 a suggestion was made if we could try something different. 

So, the idea would be to reuse the `RestServiceException` instead of creating `ServiceResponse`. The `callService` would go back to being return `void` but we'd instead throw a `RestServiceException` when things go wrong. The `RestServiceException` will need to be modified to pass an `int` like your `ServiceResponse` class.

What are the benefits of this approach? Well, the change is to throw an exception instead of returning a HTTP code when something goes wrong. Some other benefits is because `RestServiceException` is a `RuntimeException` all serverlets can remain unchanged (i.e they don't need to declare `throws RestServiceException`). It will also mean that the original functionality of when there are no exceptions everything is assumed to be a HTTP 200 so that you don't need to add the return call in all of those serverlet.

It would mean that the change will be changing perhaps 4 files rather than 25 files. Noted that you have already made the change but I would like to propose @algol60's solution too.

If you are happy to continue working on this PR and make the above change then your welcome to. If however you don't feel like doing this suggestion then that is totally fine; I'm happy to do it and put a PR for you to see.

Auriga2 commented:
No worries @arcturus, but I cannot figure out how throw a specific number error with that because the super (RuntimeException) doesn't support that. Happy to improve it if you can give me a clue.

arcturus2 commented:
Hi @Auriga2 you could do it like this:

```java
public class RestServiceException extends RuntimeException {

    private final int code;
    
    public RestServiceException(final int code, final String message) {
        super(message);
        this.code = code;
    }
    
    public RestServiceException(final String message) {
        super(message);
        this.code = HttpServletResponse.SC_INTERNAL_SERVER_ERROR;
    }

    public RestServiceException(final Exception ex) {
        super(ex);
        this.code = HttpServletResponse.SC_INTERNAL_SERVER_ERROR;
    }
    
    public int getHttpCode(){
        return this.code;
    }
}
```
In the servelets when you throw the exception using the one allowing you to pass the `int`.

Then in `ConstellationApiServlet` in the `doGet` and `doPut` you can add the extra catch for the `RestServiceException`

```java
            try {
                get(request, response);
            } catch (final RestServiceException ex) {
                response.reset();
                response.sendError(ex.getHttpCode(), ex.getMessage());
            } catch (final IOException | ServletException ex) {
                response.reset();
                response.sendError(HttpServletResponse.SC_INTERNAL_SERVER_ERROR, ex.getMessage());
                LOGGER.log(Level.INFO, "in doGet", ex);
            }
```

Thanks to @algol60 for the suggestion!

Auriga2 commented:
Thanks for the suggestion @algol60 and @arcturus2. Yes it's much better.
Let me know when this is reviewed, I can apply it to other endpoints as well. (Or create a case to fix later)

arcturus2 commented:
@Auriga2 thanks for the update

arcturus2 commented:
> Let me know when this is reviewed, I can apply it to other endpoints as well. (Or create a case to fix later)

@Auriga2 I've reviewed it. Which endpoints are left to do?

Auriga2 commented:
@arcturus2 For instance `get_attributes` throws `NullPointerException` for invalid `graph id`:
![image](https://user-images.githubusercontent.com/64826928/85819955-5f5abd80-b7b4-11ea-96a0-1e371d1d6d61.png)

Other endpoints would still throw 500 for invalid user input.   
E.g.: `Open Graph`:
![image](https://user-images.githubusercontent.com/64826928/85819321-78626f00-b7b2-11ea-9f0b-6eb4cd0f857c.png)

arcturus2 commented:
Thanks understood @Auriga2. I'm OK with either option; keep working on it in this PR or a new issue + PR. Let me know and based on this we can either merge this PR in or not.

Auriga2 commented:
Following endpoints do not show the message because of missing `Response body`. I couldn't figure out the cause. 
Get_graph_image
Get_icon
Set_graph

![image](https://user-images.githubusercontent.com/64826928/86211917-44be8500-bbb6-11ea-83c6-a1d6f8ce883a.png)

@algol60  Do you know how to fix that? I feel like it'd be nice if we can show the failure reason to the user.

Auriga2 commented:
get_plugin_description is throwing nullpointer exception for any valid plugin with no parameters. E.g.: SELECTALL, UNDIMALL, CONTRACTGRAPH
![image](https://user-images.githubusercontent.com/64826928/86314141-04b6db00-bc66-11ea-80e5-b18f8ff673e4.png)

 Plugins with parameters E.g.: RESOURCEALLOCATIONINDEXANALYTIC, STRUCTUREDGRAPHBUILDER work.

Fixed that also here.

Auriga2 commented:
I didn't add a check for missing parameters. Feels like they should also be marked as `*required` similar to the `CONSTELLATION secret`
![image](https://user-images.githubusercontent.com/64826928/86327634-c54ab780-bc82-11ea-8cc6-df3f84fdb7bf.png)
May be we can create a low priority case to fix later.

arcturus2 commented:
>Following endpoints do not show the message because of missing Response body. I couldn't figure out the cause.
Get_graph_image
Get_icon
Set_graph

I couldn't figure it out either but what you've done is fantastic.

> I didn't add a check for missing parameters. Feels like they should also be marked as *required similar to the CONSTELLATION secret

In the `SwaggerServlet` I saw
```
param.put("required", false); // TODO Hard-code this until PluginParameters grows a required field.
```
so once we implement the required property for plugins we should be able to get this automatically in Swagger.

Auriga2 commented:
Created https://github.com/constellation-app/constellation/issues/633 to fix the above issue to mark as *required.
